File: Sheets/src/neetcode/practice/Greedy/_5_handOfStraights.java
package neetcode.practice.Greedy;

import java.util.TreeMap;

public class _5_handOfStraights {
  public static boolean isNStraightHand(int[] hand, int groupSize) {
    TreeMap<Integer, Integer> map = new TreeMap<>();
    for (int num : hand) map.put(num, map.getOrDefault(num, 0) + 1);

    while (!map.isEmpty()) {
      int i = 1, currNum = map.firstKey();
      int freq = map.get(currNum);
      if (freq == 1) map.remove(currNum);
      else map.put(currNum, --freq);

      while (i++ < groupSize) {
        currNum++;
        freq = map.getOrDefault(currNum, -1);
        switch (freq) {
          case -1:
            return false;

          case 1:
            map.remove(currNum);
            break;

          default:
            map.put(currNum, --freq);
        }
      }
    }
    return true;
  }
}

File: Sheets/src/neetcode/practice/Greedy/_5_HandOfStraights.cc
#include <set>
#include <vector>

class Solution
{
public:
  bool isNStraightHand(std::vector<int> &hand, int groupSize)
  {
    std::multiset<int> set(hand.begin(), hand.end());
    if (set.size() % groupSize != 0)
      return false;

    while (!set.empty())
    {
      int size{1};
      int current = *set.begin();
      set.erase(set.begin());
      for (; size < groupSize; size++)
      {
        auto next = set.find(current + 1);
        if (next == set.end())
          return false;

        current = *next;
        set.erase(next);
      }
    }

    return true;
  }
};
